File: packages/cli/commands/doctor.ts
import { type DockerEndpoint, EngineClient } from "@compositz/core";
import { bold, dim, green, red } from "@std/fmt/colors";

/** Health check: resolve the endpoint, ping the engine, print versions. */
export async function doctor(_args: string[]): Promise<number> {
  const client = new EngineClient();
  console.log(bold("compositz doctor"));
  console.log(dim(`  endpoint: ${describeEndpoint(client.endpoint)}`));

  try {
    const pong = await client.ping();
    console.log(`  ping:     ${green(pong)}`);
    const v = await client.version();
    console.log(
      `  engine:   Docker ${v.Version}  (API ${v.ApiVersion}, min ${v.MinAPIVersion ?? "?"})`,
    );
    console.log(`  platform: ${v.Os}/${v.Arch}`);
    console.log(green("OK — engine reachable."));
    return 0;
  } catch (e) {
    console.log(red(`FAILED — ${e instanceof Error ? e.message : String(e)}`));
    console.log(dim("  Is Docker running? On Windows, Docker Desktop must be started."));
    return 1;
  }
}

function describeEndpoint(ep: DockerEndpoint): string {
  switch (ep.kind) {
    case "unix":
      return `unix://${ep.path}`;
    case "npipe":
      return `npipe://${ep.path}`;
    case "tcp":
      return `tcp://${ep.host}:${ep.port}`;
  }
}

File: crates/cli/src/commands/doctor.rs
//! `doctor` — resolve the endpoint, ping the engine, print versions.

use anyhow::Result;
use compositz_core::{connect, resolved_endpoint_description};

use crate::style::{bold, dim, green, red};

/// Health check: describe the endpoint, ping, print the engine version + platform.
/// Returns 1 (not an error) when the engine is unreachable, so the top-level
/// handler doesn't double-print — the diagnostic owns its failure message.
///
/// The header + endpoint line print BEFORE connecting: `connect()` is not lazy for
/// a unix socket (bollard checks the path exists eagerly), so a "Docker not
/// running" failure surfaces at `connect`, not at `ping`. Deriving the endpoint
/// label from the env (not a live handle) and folding a connect error into the same
/// FAILED path is what keeps the friendly diagnostic intact in exactly the case
/// `doctor` exists for.
pub async fn run() -> Result<i32> {
    println!("{}", bold("compositz doctor"));
    println!(
        "{}",
        dim(&format!("  endpoint: {}", resolved_endpoint_description()))
    );

    match probe().await {
        Ok(()) => Ok(0),
        Err(err) => {
            println!("{}", red(&format!("FAILED — {err}")));
            println!(
                "{}",
                dim("  Is Docker running? On Windows, Docker Desktop must be started.")
            );
            Ok(1)
        }
    }
}

/// The reachable-engine path: connect + ping + version, each line printed as it
/// resolves. Any failure (connect or request) propagates to the FAILED branch.
async fn probe() -> Result<()> {
    let engine = connect()?;

    let pong = engine.ping().await?;
    println!("  ping:     {}", green(&pong));

    let version = engine.version().await?;
    let field = |value: Option<String>| value.unwrap_or_else(|| "?".to_string());
    println!(
        "  engine:   Docker {}  (API {}, min {})",
        field(version.version),
        field(version.api_version),
        field(version.min_api_version),
    );
    println!(
        "  platform: {}/{}",
        version.os.unwrap_or_default(),
        version.arch.unwrap_or_default()
    );
    println!("{}", green("OK — engine reachable."));
    Ok(())
}
